ignite/pkg/jsonfile/jsonfile.go

```diff
@@ -0,0 +1,311 @@
+	// TODO create a cache system to avoid download genesis with the same hash again
+
+	// Download the file from URL
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusNotFound {
+		return nil, ErrInvalidURL
+	}
+
+	// Remove the old file if exists and create a new one
+	if err := os.RemoveAll(path); err != nil {
+		return nil, err
+	}
+	file, err := os.OpenFile(path, os.O_RDWR|os.O_APPEND|os.O_CREATE, 0600)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create the file")
+	}
+
+	// Copy the downloaded file to buffer and the opened file
+	var buf bytes.Buffer
+	if _, err := io.Copy(file, io.TeeReader(resp.Body, &buf)); err != nil {
+		return nil, err
+	}
+
+	// Check if the downloaded file is a tarball and extract only the necessary JSON file
+	var ext bytes.Buffer
+	tarballPath, err := tarball.ExtractFile(&buf, &ext, tarballFileName)
+	if err != nil && err != tarball.ErrNotGzipType && err != tarball.ErrInvalidFileName {
+		return nil, err
+	} else if err == nil {
+		// Erase the tarball bite code from the file and copy the correct one
+		if err := truncate(file, 0); err != nil {
+			return nil, err
+		}
+		if _, err := io.Copy(file, &ext); err != nil {
+			return nil, err
+		}
+	}
+
+	return &JSONFile{
+		updates:     make(map[string][]byte),
+		file:        file,
+		url:         url,
+		tarballPath: tarballPath,
+	}, nil
+}
+
+// Bytes returns the jsonfile byte array.
+func (f *JSONFile) Bytes() ([]byte, error) {
+	file := f.cache
+	if file != nil {
+		return file, nil
+	}
+	if err := f.Reset(); err != nil {
+		return nil, err
+	}
+	scanner := bufio.NewScanner(f.file)
+	for scanner.Scan() {
+		file = append(file, scanner.Bytes()...)
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+	f.cache = file
+	return file, nil
+}
+
+// Field return the param by key and the position into byte slice from the file reader.
+// Key can be a path to a nested parameter eg: app_state.staking.accounts
+func (f *JSONFile) Field(key string, param interface{}) error {
+	file, err := f.Bytes()
+	if err != nil {
+		return err
+	}
+
+	value, dataType, _, err := jsonparser.Get(file, strings.Split(key, keySeparator)...)
+	if err == jsonparser.KeyPathNotFoundError {
+		return ErrFieldNotFound
+	} else if err != nil {
+		return err
+	}
+
+	switch dataType {
+	case jsonparser.Boolean, jsonparser.Array, jsonparser.Number, jsonparser.Object:
+		err := json.Unmarshal(value, param)
+		if _, ok := err.(*json.UnmarshalTypeError); ok {
+			return ErrInvalidValueType
+		} else if err != nil {
+			return err
+		}
+	case jsonparser.String:
+		paramStr, ok := param.(*string)
+		if !ok {
+			return ErrInvalidValueType
+		}
+		*paramStr, err = jsonparser.ParseString(value)
```

Comment:
shouldn't we set `paramStr` to `param`? looks like it's not used?

Reply:
we cannot set it directly to `param` without cast to a string pointer first